### DISCUSSION TITLE
Custom firmware for a specific device/template

### DISCUSSION THREAD

**Discussion by zeebak** [arendst/Tasmota#10882] [Q&A] [answered]
I have flashed esp8266 with tasmota. It needs to be configured every time it resets.
I want the specific template to be auto configured (OR the tasmota firmware to know only the one template/config) after it resets. How do i achieve this?
Please  help!

**Comment by sfromis** [comment, marked as answer]
Well, if your device is resetting (to the point of losing config) without you controlling it, that is an issue to be mitigated. Sometimes SetOption65 helps. You can also configure a binary to include USER_TEMPLATE

**Reply by zeebak** [reply]
Pardon my English. 
I meant, If I ever want to reset the device myself and I don't want to lose the configurations except the wifi credentials.
In other words, I want to press the reset button and want to WifiConfig2, without changing the templates/configurations.

I wonder if setting a Rule to DO WifiConfig2 whenn i press the button is possible,
or maybe I could compile a code which only has the one template for my board

**Reply by sfromis** [reply]
As I wrote, you can configure template at compile time via the USER_TEMPLATE setting. Something like this in the override file:
```
#define USER_TEMPLATE "{\"NAME\":\"Generic\",\"GPIO\":[1,1,1,1,1,1,1,1,1,1,1,1,1,1],\"FLAG\":0,\"BASE\":18}"
#define MODULE        USER_MODULE
```
Obviously, adapt the template to your needs, and remember the escapes before each "

**Reply by barbudor** [reply]
And most of the other settings, setoptions, etc ... can also be enforced in `user_config_override.h`
Then if you want to compile 1 binary per device, it's your go but not very convenient

You may also have a backend that detects that an unconfigured device comes up and need configuration.

**Reply by sfromis** [reply]
The latter is what I do

**Reply by techman83** [reply]
What backend do you use @sfromis ? `USER_TEMPLATE` is super useful + `USER_MODULE` is super useful.

**Reply by sfromis** [reply]
My backend is part Node-Red, but here a homebrew setup with my own little programming language. If a new device is has a basic configuration allowing it to connect to MQTT, I can check the incoming messages, and detect a device in need of a config from my simple configuration management setup, using keys like host name or MAC address.

**Reply by techman83** [reply]
Neat, that's what I was going to do. Especially for the 20+ downlights, I messed up a firmware and reset themselves during an upgrade. Which was a minor pain :sweat_smile:

**Reply by zeebak** [reply]
Thanks. USER_TEMPLATE is super useful indeed!!
